crypto/src/main/kotlin/net/corda/v5/crypto/merkle/MerkleTreeFactory.kt

```diff
@@ -0,0 +1,41 @@
+package net.corda.v5.crypto.merkle
+
+import net.corda.v5.crypto.DigestAlgorithmName
+import net.corda.v5.base.annotations.DoNotImplement
+import net.corda.v5.base.annotations.Suspendable
+
+/**
+ * [MerkleTreeFactory] creates [MerkleTree]s, [MerkleProof]s and [MerkleTreeHashDigestProvider].
+ */
+@DoNotImplement
+interface MerkleTreeFactory {
+    /**
+     * Creates a [MerkleTree]
+     *
+     * @param leaves The leaves of the tree.
+     * @param digestProvider Merkle Tree Hash digest provider used to construct the tree's node and leaf hashes.
+     *
+     * @return A new [MerkleTree] instance.
+     */
+    @Suspendable
+    fun createTree(
+        leaves: List<ByteArray>,
+        digestProvider: MerkleTreeHashDigestProvider
+    ) : MerkleTree
+
+    /**
+     * Creates a [MerkleTreeHashDigestProvider].
+     *
+     * @param merkleTreeHashDigestProviderName name of the Hash Digest Provider class
+     * @param digestAlgorithmName name of the base Hash algorithm
+     * @param options Hash Digest provider specific options
+     *
+     * @return A new [MerkleTreeHashDigestProvider] instance.
+     */
+    @Suspendable
+    fun createHashDigestProvider(
+        merkleTreeHashDigestProviderName: MerkleTreeHashDigestProviderName,
+        digestAlgorithmName: DigestAlgorithmName,
```

Comment:
I am not sure the DigestAlgorithmName, or MerkleTreeHashDigestProviderOption are really adding much over a string. What is really needed is an explanation of what the different default digest providers do and what parameters they require. This would be best of the platform defined constants were grouped next to each other rather than in separate files and it does need considerably more text.

Reply:
If we do go with MerkleTreeHashDigestProviderOption as a distinct class it is of value to include a property to tell you the expected value type that would go into the options map